subschemas/substrate-chain/src/resolvers/Query/parathreads.ts

```diff
@@ -0,0 +1,182 @@
+
+type ParaMap = {
+  id: ParaId;
+  leases: LeaseInfo[];
+};
+
+export type PartialManager = Omit<Manager, 'account'>;
+
+interface ParathreadData extends Omit<Parathread, 'manager'> {
+  manager: PartialManager;
+}
+
+export async function parathreads(
+  _: Record<string, never>,
+  __: Record<string, never>,
+  {api}: Context,
+): Promise<ParathreadData[]> {
+  const paraLifecycles = (await api.query.paras?.paraLifecycles?.entries()) as ParaIdEntries;
+  const parathreadIds = paraLifecycles ? extractIds(paraLifecycles) : [];
+  const genesisHash = api.genesisHash.toHex();
+  const wsEndpoints = createWsEndpoints((key: string, value: string | undefined) => value || key);
+  const endpoints = wsEndpoints.filter(({genesisHashRelay}) => genesisHash === genesisHashRelay);
+  const hasLinksMap = isParasLinked(parathreadIds, endpoints);
+  const leases = (await api.query.slots.leases.multi(parathreadIds)) as unknown as LeaseOptions[];
+  const paraMap = extractParaMap(hasLinksMap, parathreadIds, leases);
+  const length = api.consts.slots.leasePeriod as BlockNumber;
+  const bestNumber = await api.derive.chain.bestNumber();
+  const currentPeriod = bestNumber.div(length);
+
+  const parathreads = Promise.all(
+    paraMap.map(async (para) => {
+      const optInfo = (await api.query.registrar.paras(para.id)) as Option<PolkadotRuntimeCommonParasRegistrarParaInfo>;
+      const parathreadLink = endpoints.find((endpoint) => endpoint.paraId?.toString() === para.id.toString());
+      const progress = bestNumber.mod(length);
+      const leasePeriod = {
+        currentPeriod,
+        length,
+        progress,
+        remainder: length.sub(progress),
+      };
+      const {periodString, blocks} = parseLeases({leasePeriod, leases: para.leases});
+
+      return {
+        id: para.id.toString(),
+        manager: {address: optInfo.isSome ? optInfo.unwrap().manager.toString() : ''},
+        name: parathreadLink?.text ?? null,
+        homepage: parathreadLink?.homepage ?? null,
+        lease: {
+          period: periodString,
+          blockTime: getBlockTime(api, blocks).timeStringParts,
+        },
+      };
+    }),
+  );
+
+  return parathreads;
+}
+
+function isParasLinked(ids: ParaId[], endpoints: LinkOption[]) {
+  return ids.reduce(
+    (all: Record<string, boolean>, id) => ({
+      ...all,
+      [id.toString()]: extractParaEndpoints(endpoints, id).length !== 0,
+    }),
+    {},
+  );
+}
+
+function parseLeases({leases, leasePeriod}: {leases: LeaseInfo[]; leasePeriod: LeasePeriod}) {
+  const leasesPeriods = leases?.map((lease) => lease?.period).filter(notEmpty);
+  const periodString = getLeasePeriodString(leasePeriod.currentPeriod, leasesPeriods);
+
+  const periods = leasePeriod?.currentPeriod ? leasesPeriods : undefined;
+  const firstPeriod = periods?.[0];
+
+  const blocks =
+    leasePeriod && firstPeriod && bnToBn(firstPeriod).sub(BN_ONE).imul(leasePeriod.length).iadd(leasePeriod.remainder);
```

Comment:
Here remainder is already of type `BN`, there is no need to get the string and parse it again

Reply:
Yeah, true. I was still thinking in terms of the `getLeasePeriod` function! Was not about the `bnToBn` but rather which field to use. Since you construct your own leasePeriod, it's good - sorry for that